cpp/src/arrow/util/mutex.h

```diff
@@ -0,0 +1,58 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+
+#include "arrow/util/macros.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+namespace util {
+
+/// A wrapper around std::mutex since we can't use it directly due to CPP/CLI
```

Comment:
we use `std:mutex` in several places in the code base, so this comment is not fully clear to me

Reply:
We cannot use `std::mutex` in public headers due to CPP/CLI forbidding use of that header https://docs.microsoft.com/en-us/cpp/standard-library/mutex#remarks
We don't support compilation of Arrow with the `/clr` option so it's acceptable to use `std::mutex` in source files and private headers (including `arrow/util/mutex.cc`). However arrow headers may be included by projects which *are* using `/clr` and usage of `std::mutex` in those public header files would break those projects.

Reply:
Thanks for the explanation! maybe add: "can't use it directly *in public headers*"